Avoid printing candidates in resolution error messages that represent a method / standalone function mismatch

bradcray opened the pull request:
This implements an idea proposed in issue #16928 in which we don't print out resolution candidates by default if they are standalone functions and the call is a method; or the call is to a standalone function and the candidate is a method.  This was motivated by a user error reported in issue #16917 in which a bad call to `write()` printed out four unrelated method-based implementations of `write()`.

As an example, prior to this change, calling `write("hi", int);` resulted in:

```bash
badWrite.chpl:1: error: unresolved call 'write("hi", type int(64))'
$CHPL_HOME/modules/standard/ChapelIO.chpl:837: note: this candidate did not match: write(const args ...?n)
badWrite.chpl:1: note: because actual argument #2 is a type
$CHPL_HOME/modules/standard/ChapelIO.chpl:837: note: but is passed to non-type varargs formal 'args'
$CHPL_HOME/modules/internal/Atomics.chpl:237: note: candidates are: AtomicBool.write(value: bool, param order: memoryOrder = memoryOrder.seqCst)
$CHPL_HOME/modules/internal/Atomics.chpl:422: note:                 AtomicT.write(value: T, param order: memoryOrder = memoryOrder.seqCst)
note: and 2 other candidates, use --print-all-candidates to see them
```

where this PR changes it to:

```bash
badWrite.chpl:1: error: unresolved call 'write("hi", type int(64))'
$CHPL_HOME/modules/standard/ChapelIO.chpl:837: note: this candidate did not match: write(const args ...?n)
badWrite.chpl:1: note: because actual argument #2 is a type
$CHPL_HOME/modules/standard/ChapelIO.chpl:837: note: but is passed to non-type varargs formal 'args'
note: there are also 4 other candidates, use --print-all-candidates to see them
```

Resolves #16928 to my satisfaction (that issue also proposed that we trim method calls to classes not in the same class sub-hierarchy, but I currently think that this happens rarely enough to not be worth the additional effort until it becomes a true pain point).

One interesting side effect of this is that in some existing tests, the "best" match whose failed resolution we choose to describe sometimes changed between a standalone function and a method.  For example, given `foo(1)` and choices between `proc R.foo(x:int)` and `proc foo()`, perhaps we used to choose the method and describe the missing receiver, but now choose the standalone function and describe the extraneous argument.  This is arguably a "half-empty/half-full" type of issue where there's arguably no right answer (e.g., did the programmer forget to mention the object on which they intended to dispatch? or did they think there was an int-taking standalone version that doesn't exist?).  But I think the benefits are sufficiently strong in cases like the one motivating that this tradeoff is worth taking as well.

Added two new tests:
* one that captures a version of the original motivating issue
* one that tests a number of method vs. other class method vs. standalone function overload errors to make sure they behaved as expected

While here, I made some other minor adjustments:

* changed "1 arguments" to "1 argument" in a few error messages because it started to bug me
* changed "candidates are:" line after an explanation about the first bad match to "other candidates include" because the previous phrasing seemed confusing (since it wasn't listing all candidates, simply those that hadn't been explained)
* moved these "other candidates" to their own lines with no indentation rather than indenting them as before, as suggested by @mppf and @vasslitvinov 
* specialized the message when there are no other candidates that meet the standalone/method choice to say something like "there are also 4 other candidates" to make it read more clearly (to my eyes)
* refactored some of the compiler logic that generates these error messages in hopes of making it a bit more straightforward

TODO:
- [x] remove dead code introduced in this PR

vasslitvinov commented:
Nice improvement!

Looking at the new text in `test/classes/deitz/method/method_call2.good `

> note: this candidate did not match: bar.foo()
> note: because call includes 1 argument
> note: but function can only accept 0 arguments

The first line talks about a "candidate" whereas the third line says "function". This leaves me wondering whether those two are the same or different. Can both lines use "candidate"?

I don't like "only" coming before "accept". It needs to come before the thing that it is qualifying. Otherwise it sounds like you are saying "only accept" so it cannot, for example, "reject". I'd like to see it say "function can accept only 0 arguments". This is not entirely perfect, yet an improvement.

The previous .good says:

> note: other candidates include:
> note: D2.x

This makes me think that there are more "other candidates" that are not shown here. Can it instead  say "[the] other candidates are:" ? That would make it definitive.

mppf commented on test/classes/deitz/method/method_call2.good:
```diff
@@ -1,6 +1,6 @@
+method_call2.chpl:6: note: this candidate did not match: bar.foo()
+method_call2.chpl:10: note: because call includes 1 argument
+method_call2.chpl:6: note: but function can only accept 0 arguments
+note: there is also 1 other candidate, use --print-all-candidates to see them
```
"to see them" -> "to see it" if there's just one?

bradcray replied:
D'oh!

mppf commented on compiler/resolution/functionResolution.cpp:
```diff
@@ -4286,18 +4301,41 @@ static void generateUnresolvedMsg(CallInfo& info, Vec<FnSymbol*>& visibleFns) {
+    // If there is more than one visible function and we haven't been
+    // asked to print all candidates, let's try to eliminate "worse"
+    // matches (e.g., method call / standalone function call
+    // mismatches).
```
Are you sure we don't want to just adjust the sorting used above in `sortExampleCandidates` ? If you can arrange for the candidates in the case you care about to appear first in the sorted list, that might work better than filtering. It has slightly different behavior but it seems better IMO.

bradcray replied:
In the original motivating example, the problem wasn't the order in which the unrelated `.write()` methods were printed, it was more that they were printed at all, which is what this PR is trying to improve.  So I don't see that adjusting the sort order would help without these other changes?

mppf replied:
That's fine; I haven't been following the motivating example in detail.

bradcray replied:
It's in the OP if you're interested.  The idea is that none of the other 4 candidates for write() are printed because none of them are standalone as the function call itself was.

bradcray commented:
@vasslitvinov:

> The first line talks about a "candidate" whereas the third line says "function". This leaves me wondering whether those two are the same or different. Can both lines use "candidate"?

> I don't like "only" coming before "accept". It needs to come before the thing that it is qualifying. Otherwise it sounds like you are saying "only accept" so it cannot, for example, "reject". I'd like to see it say "function can accept only 0 arguments". This is not entirely perfect, yet an improvement.

These are both pre-existing phrases that are not new as a result of this PR; the reason they seem new in the test you cite is that previously another candidate had been described.  But if you `git grep` your copy of the `test` directory, you'll see other tests that include those phrasings today.  They may be worthy of further discussion and improvement, but I'm not interestedin taking that on here (arguably I've already changed more than I should've in one PR, but it was difficult for me to introduce new grammatical errors, which is why I fixed a few others).

> Can it instead say "[the] other candidates are:" ? That would make it definitive.

Yeah, I could change that back.

bradcray commented:
@e-kayrakli has also proposed restoring a modest amount of indentation, but not all of it, for readability.  So rather than the previous approach of heavy indentation:

```bash
testBadCallCandidates.chpl:20: note: other candidates are:
testBadCallCandidates.chpl:8: note:                        C.bar(x: string)
testBadCallCandidates.chpl:11: note:                       C.bar(x: D)
testBadCallCandidates.chpl:65: note:                       bar(x: D)
```

or the PR's current approach of no indentation:

```bash
testBadCallCandidates.chpl:20: note: other candidates are:
testBadCallCandidates.chpl:8: note: C.bar(x: string)
testBadCallCandidates.chpl:11: note: C.bar(x: D)
testBadCallCandidates.chpl:65: note: bar(x: D)
```

we might instead use a modest amount of indentation, say:

```bash
testBadCallCandidates.chpl:20: note: other candidates are:
testBadCallCandidates.chpl:8: note:   C.bar(x: string)
testBadCallCandidates.chpl:11: note:   C.bar(x: D)
testBadCallCandidates.chpl:65: note:   bar(x: D)
```

Did I get that right, Engin?

[there are obvious additional improvements we could make like removing the forced `note:` on every line, which seems more annoying than helpful and potentially making the notes and errors a bit more independent of the number of columns the line number takes up (restoring alignment in the case just above), but that's above and beyond what I'm trying to do here.

e-kayrakli commented:
> Did I get that right, Engin?

Yes that looks much better, thanks!

bradcray commented:
@mppf: Would you be up for reviewing this since it messes with code you developed?  I think it's ready now.

mppf commented on compiler/resolution/functionResolution.cpp:
```diff
@@ -4228,6 +4228,21 @@ static bool isInitEqualsPreResolve(CallExpr* call) {
 }
 
+static bool obviousMisMatch(CallExpr* call, FnSymbol* fn) {
+  bool isMethod = isMethodPreResolve(call);
```
I think this function would be more clearly written

```
bool isMethodCall = isMethodPreResolve(call);
bool isMethodFn = (fn->_this != NULL);
return (isMethodCall != isMethodFn);
```

bradcray replied:
Yeah, I agree (and realized this after I requested the review, but figured I'd save all the changes for one pass).  I wrote it this way originally because (a) I didn't think it would be this simple and (b) I had aspirations to analyze the class hierarchies that a method call was involved in to eliminate unrelated class hierarchies.  But then I was wrong on the first and didn't bother with the second.

mppf commented on compiler/resolution/functionResolution.cpp:
```diff
@@ -4286,18 +4301,41 @@ static void generateUnresolvedMsg(CallInfo& info, Vec<FnSymbol*>& visibleFns) {
+    // candidates.
+    if (filteredFns.n == 0) {
+      filteredFns = visibleFns;
+    }
```
I think it would make the below code simpler / easier to follow if we limit filteredFns to `nPrint` candidates here. You could rename the variable `printFns`. I wouldn't expect this to require you to change any `.good` files. What you have is OK if you don't want to pursue this.

bradcray replied:
To make sure I'm understanding, the idea is:

* once we reach this line, `filteredFns` will always have `nPrint` or fewer candidates because that's all we'll add to it
* then the logic below gets simplified because we never hit the `i > nPrint` condition and don't have to worry about the `i <= nPrint` condition.  In essence, we print "all" the filtered functions and then have a simpler condition for printing out the "other candidates" addendum

Did I get that right?  (I found that logic pretty squirrelly, so like the idea of simplifying it)

bradcray replied:
Took a stab at this in https://github.com/chapel-lang/chapel/pull/16969/commits/a60e22fc990a0fda59e97b7e005fb23a09eb6e00.  Running a full test now.

bradcray replied:
Looks like I didn't get something quite right.  Will take another look tomorrow.

mppf replied:
Yes that is what I was proposing.

mppf reviewed (approved):
Thanks for improving the error messages!

mppf commented on compiler/resolution/functionResolution.cpp:
```diff
@@ -4297,32 +4297,41 @@ static void generateUnresolvedMsg(CallInfo& info, Vec<FnSymbol*>& visibleFns) {
+      // asked to print all candidates, let's try to eliminate "worse"
+      // matches (e.g., method call / standalone function call
+      // mismatches).
+      if (visibleFns.n > 1) {
```
Maybe the issue is that you meant for this condition to be combined with 4304? The way it is now, I think it would put nothing in filteredFns if there is 1 candidate, and then in the logic below, add all of the visible functions.

Anyway I think it would be nicer to put the logic about filteredFns.n == 0 in this else block rather than in another condition below.

I'm thinking

```
 if visibleFns.n <= 1 || fPrintAllCandidates
   filteredFns = visibleFns;
 else
   add not-obvious-mismatches to filteredFns
   if filteredFns is empty, copy over the first nPrint
 ```
 
 but of course there are lots of other ways to write it

mppf replied:
I take that back, I don't think that's the issue, but you can adjust the logic to my suggestion if you like it better.

bradcray replied:
It was more like an off-by-one error, but fixing it gave me the motivation to do some other cleanups that had been in the back of m mind.  I also ended up doing something like your suggestion here, which is a bit more 6 vs. 1/2 dozen to me but definitely makes the n==1 case less convoluted...  Re-testing now.

bradcray replied:
@mppf: This is back to testing cleanly and is fairly improved in terms of its refactoring, I think.  Would you mind taking another look to see if you agree before I merge it?  Thanks!

mppf reviewed (approved):
Looks great to me! Thanks for cleaning up the twisted logic.
